src/ray/rpc/client_call.h

```diff
@@ -0,0 +1,165 @@
+#ifndef RAY_RPC_CLIENT_CALL_H
+#define RAY_RPC_CLIENT_CALL_H
+
+#include <grpcpp/grpcpp.h>
+#include <boost/asio.hpp>
+
+#include "ray/common/status.h"
+#include "ray/rpc/util.h"
+
+namespace ray {
+namespace rpc {
+
+/// Represents an outgoing gRPC request.
+///
+/// The lifecycle of a `ClientCall` is as follows.
+///
+/// When a client submits a new gRPC request, a new `ClientCall` object will be created
+/// by `ClientCallMangager::CreateCall`. Then the object will be used as the tag of
+/// `CompletionQueue`.
+///
+/// When the reply is received, `ClientCallMangager` will get the address of this object
+/// via `CompletionQueue`'s tag. And the manager should call `OnReplyReceived` and then
+/// delete this object.
+class ClientCall {
+ public:
+  /// The callback to be called by `ClientCallManager` when the reply of this request is
+  /// received.
+  virtual void OnReplyReceived() = 0;
+};
+
+class ClientCallManager;
+
+/// Reprents the client callback function of a particular rpc method.
+///
+/// \tparam Reply Type of the reply message.
+template <class Reply>
+using ClientCallback = std::function<void(const Status &status, const Reply &reply)>;
+
+/// Implementaion of the `ClientCall`. It represents a `ClientCall` for a particular
+/// RPC method.
+///
+/// \tparam Reply Type of the Reply message.
+template <class Reply>
+class ClientCallImpl : public ClientCall {
+ public:
+  void OnReplyReceived() override { callback_(GrpcStatusToRayStatus(status_), reply_); }
+
+ private:
+  /// Constructor.
+  ///
+  /// \param[in] callback The callback function to handle the reply.
+  ClientCallImpl(const ClientCallback<Reply> &callback) : callback_(callback) {}
+
+  /// The reply message.
+  Reply reply_;
+
+  /// The callback function to handle the reply.
+  ClientCallback<Reply> callback_;
+
+  /// The response reader.
+  std::unique_ptr<::grpc::ClientAsyncResponseReader<Reply>> response_reader_;
```

Comment:
Do we need `::grpc::` everywhere or would `grpc::` suffice?

Reply:
using prepended `::` means that resolution occurs from the global namespace, see https://stackoverflow.com/a/4269064.
`grpc::` should also work. I just feel `::grpc::` is more clear.

Reply:
Hmmmm, I'd suggest we get rid of it, for common package names like `std`, `arrow` or `grpc` it's usually common to not prefix it with `::` (or at least I have never seen it). What do you think?

Reply:
I'm fine with either way. Will remove this.